src/iosched/unified.c

```diff
@@ -1991,104 +2253,115 @@ int _unified_flush_all(struct unified_data *priv)
 {
 	int ret;
 	struct dentry_priv *dpr, *aux;
+	bool empty = false;
 
 	CHECK_ARG_NULL(priv, -LTFS_NULL_ARG);
 
+	ltfs_thread_mutex_lock(&priv->writer_lock);
 	acquirewrite_mrsw(&priv->lock);
 
 	if (! TAILQ_EMPTY(&priv->dp_queue)) {
 		TAILQ_FOREACH_SAFE(dpr, &priv->dp_queue, dp_queue, aux) {
+			ltfsmsg(LTFS_DEBUG, 13033D, "DP", dpr->dentry->platform_safe_name);
+			ltfs_mutex_lock(&dpr->dentry->iosched_lock);
 			ret = _unified_flush_unlocked(dpr->dentry, priv);
+			ltfs_mutex_unlock(&dpr->dentry->iosched_lock);
 			if (ret < 0) {
 				ltfsmsg(LTFS_ERR, 13020E, dpr->dentry->platform_safe_name, ret);
 				releasewrite_mrsw(&priv->lock);
 				return ret;
 			}
 		}
+	} else {
+		empty = true;
 	}
 
 	if (! TAILQ_EMPTY(&priv->working_set)) {
 		TAILQ_FOREACH_SAFE(dpr, &priv->working_set, working_set, aux) {
+			ltfsmsg(LTFS_DEBUG, 13033D, "WS", dpr->dentry->platform_safe_name);
+			ltfs_mutex_lock(&dpr->dentry->iosched_lock);
 			ret = _unified_flush_unlocked(dpr->dentry, priv);
+			ltfs_mutex_unlock(&dpr->dentry->iosched_lock);
 			if (ret < 0) {
 				ltfsmsg(LTFS_ERR, 13020E, dpr->dentry->platform_safe_name, ret);
 				releasewrite_mrsw(&priv->lock);
 				return ret;
 			}
 		}
+	} else {
+		if (empty)
+			empty = true;
+		else
+			empty = false;
```

Comment:
This block below looks funny. Are you sure this is what you wanted to have here? You can probably delete it.

Reply:
I would like to set `bool empty` only when both queues are empty.

| empty flag | dp_queue  | working_set |
| -- | -- | -- | 
| true | EMPTY | EMPTY |
| false | NOT EMPTY | EMPTY |
| false | EMPTY | NOT EMPTY |
| false | NOT EMPTY | NOT EMPTY |

Reply:
How about this ?

```C
	ltfs_thread_mutex_lock(&priv->writer_lock);
	acquirewrite_mrsw(&priv->lock);

	/* First of all, test both are empty */
	if (TAILQ_EMPTY(&priv->dp_queue) && TAILQ_EMPTY(&priv->working_set)) {
		empty = true;
	} else {
		if (! TAILQ_EMPTY(&priv->dp_queue)) {
			TAILQ_FOREACH_SAFE(dpr, &priv->dp_queue, dp_queue, aux) {
				ltfsmsg(LTFS_DEBUG, 13033D, "DP", dpr->dentry->platform_safe_name);
				ltfs_mutex_lock(&dpr->dentry->iosched_lock);
				ret = _unified_flush_unlocked(dpr->dentry, priv);
				ltfs_mutex_unlock(&dpr->dentry->iosched_lock);
				if (ret < 0) {
					ltfsmsg(LTFS_ERR, 13020E, dpr->dentry->platform_safe_name, ret);
					releasewrite_mrsw(&priv->lock);
					return ret;
				}
			}
		}

		if (! TAILQ_EMPTY(&priv->working_set)) {
			TAILQ_FOREACH_SAFE(dpr, &priv->working_set, working_set, aux) {
				ltfsmsg(LTFS_DEBUG, 13033D, "WS", dpr->dentry->platform_safe_name);
				ltfs_mutex_lock(&dpr->dentry->iosched_lock);
				ret = _unified_flush_unlocked(dpr->dentry, priv);
				ltfs_mutex_unlock(&dpr->dentry->iosched_lock);
				if (ret < 0) {
					ltfsmsg(LTFS_ERR, 13020E, dpr->dentry->platform_safe_name, ret);
					releasewrite_mrsw(&priv->lock);
					return ret;
				}
			}
		}
	}

	releasewrite_mrsw(&priv->lock);
```